umbraco/Umbraco-CMS
# Netcore: Install steps in dotnet core executable

Fixes [AB#6233](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/6233) 

### Details
-  Needed to update some dictionarys in config to be case insensitive
-  Small changes to the typefinder
     - Needed to explicit add "System.Data.SqlserverCe" assembly to the list of `DefaultUmbracoAssemblyProvider.UmbracoCoreAssemblyNames` to ensure the assembly is found 
     - Using ` Assembly.LoadFrom(dll)` instead of `Assembly.Load(assemblyName)`
    - Changed `KnownAssemblyExclusionFilter` because the "System." is too generelt now that we need to find `System.Data.SqlserverCe`.
- Created a new `IDbProviderFactoryCreator ` where we inject all possible creators but find the one based on the provider name.. Input to this is very welcome.
    - Added `ProviderName` property on
         - `IBulkSqlInsertProvider`
         - `ISqlSyntaxProvider`
- Removed the version from Config and the `UmbracoVersion.LocalVersion`
- Implemented our own `HttpResponseException ` and a filter to catch these exceptions and return the expected status code.
- Turned off the `WithAspNetCoreSettings` on the container, such that we can register multiple implementations on a single interface, and then inject these as `IEnumerable<...>`
- Macros
    - Due to dependencies we needed the `MacroRenderer` to be ported. I did that, by using a ViewComponent instead of controller with child only action.
- Had to add `AppDomain.CurrentDomain.SetData("DataDirectory", IOHelper.MapPath(Constants.SystemDirectories.Data));` in CoreRuntime, to ensure the `DataDirectory` part of the embedded connectionstrings are correct. 


### Known issues
- I introduced a hack to the BTree.  There is comment in code.
- We still need to implement `MemberUserKeyProvider` for .net core
- We still need to determine the server address. Right now the first request will fail because we try to start the background tasks
- Restarts only works when using IIS/IISExpress not Kestrel directly, due to the fact that we close the app and use the host to start a new on next request.
- It will not be possible to install Umbraco on the old .NET Framework executable

### Testing
- Start the dotnet core executable using IISExpress
- Go to `/Install`
- Follow the steps
- Wait for install and restart
- Verify redirected to `/umbraco`
- Verify connection string is updated in `appsettings.json`
- Verify when restarting, the runtime state is `Run`
- Verify `/install` redirects to `/umbraco`
